### PR TITLE
State Backend Work

### PR DESCRIPTION
**Pull request by narang-amit**

### CODE REVIEW

**Comment by narang-amit** [comment]
Temporary placeholder of a PR—will add details/pass lint on pre-existing files separately.

File: ralf/tables/connector.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    @abstractmethod
+    def get_one(self, schema: Schema, key: str) -> Union[Record, None]:
+        pass
+
+    @abstractmethod
+    def get_all(self, schema: Schema) -> List[Record]:
+        pass
+
+    @abstractmethod
+    def get_num_records(self, schema: Schema) -> int:
```
Comment (simon-mo) [inline ralf/tables/connector.py:29]: get_num_records -> count

File: ralf/tables/redis_connector.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    def __init__(self, conn: redis.client.Redis):
+        self.conn = conn
+
+    def add_table(self, schema: Schema, historical: bool):
+        pass
+
+    def update(self, schema: Schema, historical: bool, record: Record):
+        key = getattr(record, schema.primary_key)
+        jsonString = json.dumps(vars(record))
```
Comment (simon-mo) [inline ralf/tables/redis_connector.py:21]: snake case

File: ralf/tables/redis_connector.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        self.conn.hset(schema.get_name(), key, jsonString)
+
+    def delete(self, schema: Schema, key: str):
+        self.conn.hdel(schema.get_name(), key)
+
+    def get_one(self, schema: Schema, key) -> Union[Record, None]:
+        val = self.conn.hget(schema.get_name(), key)
+        if val:
+            record = val.decode("utf-8")
```
Comment (simon-mo) [inline ralf/tables/redis_connector.py:30]: do we need the decode here? json.loads should be able to handle bytes

File: ralf/tables/redis_connector.py (hunk 2 above)
Comment (simon-mo) [reply ralf/tables/redis_connector.py:21]: actually should we pickle instead of json because pickle is nicer to np array and dataframe

**Review by simon-mo — COMMENTED** [review]
- remove test.db from commit
- add https://pypi.org/project/pytest-benchmark/ for verifying the benchmarking

File: ralf/tables/sqlite_connector.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        curr = self.conn.cursor()
+        table_name = schema.get_name()
+        rows = curr.execute(f"SELECT record FROM {table_name}").fetchall()
+        records = [pickle.loads(i[0]) for i in rows]
+        return records
+
+    def count(self, schema: Schema) -> int:
+        curr = self.conn.cursor()
+        schema.get_name()
+        count = curr.execute("SELECT COUNT(id) FROM {table_name}").fetch_one()
```
Comment (simon-mo) [inline ralf/tables/sqlite_connector.py:73]:
```suggestion
        count = curr.execute(f"SELECT COUNT(id) FROM {table_name}").fetch_one()
```

**Review by simon-mo — APPROVED** [review]

File: ralf/tables/sqlite_connector.py (hunk 4 above)
Comment (simon-mo) [reply ralf/tables/sqlite_connector.py]: @narang-amit rip i think this broke lint, can you help fix?

**Pull request merged**